File: src/main/java/cc/xiaobaicz/utils/error/ErrorUtil.java
package cc.xiaobaicz.utils.error;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Build;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PrintStream;
import java.io.RandomAccessFile;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * 崩溃日志
 * @author BC
 * @date 2019/8/28
 */
public final class ErrorUtil implements Thread.UncaughtExceptionHandler {

    private static volatile Thread.UncaughtExceptionHandler uncaughtExceptionHandler;

    private ErrorDB mErrorDB;

    private ErrorUtil(final Context context) {
        if (context == null)
            throw new NullPointerException();
        uncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler();
        mErrorDB = new ErrorDB(context.getApplicationContext());
    }

    @Override
    public void uncaughtException(Thread t, Throwable e) {
        onError(t, e);
        uncaughtExceptionHandler.uncaughtException(t, e);
    }

    /**
     * 初始化
     * @param context context
     */
    public static void init(Context context) {
        if (context == null)
            throw new NullPointerException();
        if (!(uncaughtExceptionHandler instanceof ErrorUtil))
            Thread.setDefaultUncaughtExceptionHandler(new ErrorUtil(context));
    }

    /**
     * 写入错误日志
     * @param t 异常线程
     * @param e 异常信息
     */
    private void onError(Thread t, Throwable e) {
        final String tName = t.getName();
        final long time = System.currentTimeMillis();
        final String model = Build.MODEL;
        final String sys = String.format("Android %s", Build.VERSION.RELEASE);

        final String type;
        if (e.getCause() != null) {
            type = e.getCause().getClass().getName();
        } else {
            type = e.getClass().getName();
        }

        final String msg;
        ByteArrayOutputStream os = new ByteArrayOutputStream(1024);
        try (PrintStream ps = new PrintStream(os)) {
            e.printStackTrace(ps);
            msg = os.toString();
        }

        try (SQLiteDatabase db = mErrorDB.getWritableDatabase()) {
            ContentValues values = new ContentValues();
            values.put("thread", tName);
            values.put("time", time);
            values.put("sys", sys);
            values.put("model", model);
            values.put("type", type);
            values.put("msg", msg);
            db.beginTransaction();
            try {
                db.insert("error_log", null, values);
                db.setTransactionSuccessful();
            } finally {
                db.endTransaction();
            }
        }
    }

    /**
     * 获取日志
     * @return 日志集
     */
    public static List<Error> getLogs(Context context, int index, int count) {
        if (context == null)
            throw new NullPointerException();
        if (index < 0 || count < 0)
            throw new IllegalStateException();
        List<Error> errors = new ArrayList<>();
        try (
                SQLiteDatabase db = new ErrorDB(context.getApplicationContext()).getReadableDatabase();
                Cursor cursor = db.query("error_log", new String[]{"thread", "time", "sys", "model", "type", "msg"}, null, null, null, null, "time desc", String.format("%s, %s", index, count))
        ) {
            Error error;
            while (cursor.moveToNext()) {
                error = new Error();
                error.thread = cursor.getString(0);
                error.time = cursor.getLong(1);
                error.sys = cursor.getString(2);
                error.model = cursor.getString(3);
                error.type = cursor.getString(4);
                error.msg = cursor.getString(5);
                errors.add(error);
            }
        }
        return errors;
    }

    /**
     * 获取所有日志
     * @return 日志集
     */
    public static List<Error> getLogs(Context context) {
        return getLogs(context, 0, Integer.MAX_VALUE);
    }

    /**
     * 获取最新日志
     * @return 最新日志
     */
    public static Error getNewLog(Context context) {
        if (context == null)
            throw new NullPointerException();
        try (
                SQLiteDatabase db = new ErrorDB(context.getApplicationContext()).getReadableDatabase();
                Cursor cursor = db.query("error_log", new String[]{"thread", "time", "sys", "model", "type", "msg"}, null, null, null, null, "time desc", "1")
        ) {
            Error error;
            while (cursor.moveToNext()) {
                error = new Error();
                error.thread = cursor.getString(0);
                error.time = cursor.getLong(1);
                error.sys = cursor.getString(2);
                error.model = cursor.getString(3);
                error.type = cursor.getString(4);
                error.msg = cursor.getString(5);
                return error;
            }
        }
        return null;
    }

    /**
     * 清空所有日志
     * @return 是否成功，false清空失败，true清空成功
     */
    public static void clear(Context context) {
        if (context == null)
            throw new NullPointerException();
        try (SQLiteDatabase db = new ErrorDB(context.getApplicationContext()).getReadableDatabase()) {
            db.delete("error_log", null, null);
        }
    }

    /**
     * 导出全部崩溃记录
     */
    public static void export(Context context, File path) throws FileNotFoundException {
        List<Error> logs = getLogs(context);
        export(path, logs);
    }

    /**
     * 导出指定崩溃记录
     */
    public static void export(File path, List<Error> errors) throws FileNotFoundException {
        if (path == null)
            throw new NullPointerException();
        if (!path.exists()) {
            if (!path.mkdirs()) {
                throw new FileNotFoundException();
            }
        }

        final StringBuilder sb = new StringBuilder();
        sb.append(String.format("%s,%s,%s,%s,%s,%s\n", "thread", "time", "sys", "model", "type", "msg"));

        for (Error log : errors) {
            sb.append(String.format(Locale.CHINA, "%s,%tF %tT,%s,%s,%s,%s\n", log.thread, log.time, log.time, log.sys, log.model, log.type, log.msg.replace("\n", " ~ ")));
        }

        final long now = System.currentTimeMillis();
        File file = new File(path, String.format(Locale.CHINA, "%tF-%tT.csv", now, now));
        try (final RandomAccessFile out = new RandomAccessFile(file, "rw")) {
            out.write(sb.toString().getBytes(Charset.forName("GB2312")));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: helper/src/main/java/cc/xiaobaicz/helper/error/ErrorUtil.kt
package cc.xiaobaicz.helper.error

import android.content.ContentValues
import android.content.Context
import android.os.Build
import java.io.*
import java.nio.charset.Charset
import java.util.*

/**
 * 崩溃日志
 * @author BC
 * @date 2019/8/28
 */
class ErrorUtil private constructor(context: Context?) : Thread.UncaughtExceptionHandler {
    private val CALLBACKS: MutableSet<ErrorCallback> = HashSet()
    private val mErrorDB: ErrorDB

    init {
        if (context == null) throw NullPointerException()
        mErrorDB = ErrorDB(context.applicationContext)
    }

    /**
     * 默认错误接收器
     */
    private val ERROR_CALLBACK = ErrorCallback { t, e ->
        val tName = t.name
        val time = System.currentTimeMillis()
        val model = Build.MODEL
        val sys = String.format("Android %s", Build.VERSION.RELEASE)
        val type: String
        type = if (e.cause != null) {
            e.cause!!.javaClass.name
        } else {
            e.javaClass.name
        }
        var msg: String
        val os = ByteArrayOutputStream(1024)
        PrintStream(os).use { ps ->
            e.printStackTrace(ps)
            msg = os.toString()
        }
        mErrorDB.writableDatabase.use { db ->
            val values = ContentValues()
            values.put("thread", tName)
            values.put("time", time)
            values.put("sys", sys)
            values.put("model", model)
            values.put("type", type)
            values.put("msg", msg)
            db.beginTransaction()
            try {
                db.insert("error_log", null, values)
                db.setTransactionSuccessful()
            } finally {
                db.endTransaction()
            }
        }
    }

    override fun uncaughtException(t: Thread, e: Throwable) {
        for (callback in CALLBACKS) {
            callback.callback(t, e)
        }
        uncaughtExceptionHandler.uncaughtException(t, e)
    }

    companion object {
        @Volatile
        private var uncaughtExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()
        private var sErrorUtil: ErrorUtil? = null

        /**
         * 初始化
         * @param context context
         */
        fun init(context: Context?) {
            if (context == null) throw NullPointerException()
            if (uncaughtExceptionHandler !is ErrorUtil) {
                if (sErrorUtil != null) {
                    sErrorUtil!!.mErrorDB.close()
                    sErrorUtil!!.CALLBACKS.clear()
                }
                Thread.setDefaultUncaughtExceptionHandler(ErrorUtil(context).also { sErrorUtil = it })
                register(sErrorUtil!!.ERROR_CALLBACK)
            }
        }

        /**
         * 注册监听
         * @param callback 监听
         */
        fun register(callback: ErrorCallback) {
            if (sErrorUtil == null) throw NullPointerException()
            sErrorUtil!!.CALLBACKS.add(callback)
        }

        /**
         * 获取指定类型日志
         * @return 日志集
         */
        fun getLogs(context: Context?, type: String, index: Int, count: Int): List<Error> {
            if (context == null) throw NullPointerException()
            check(!(index < 0 || count < 0))
            val errors: MutableList<Error> = ArrayList()
            ErrorDB(context.applicationContext).readableDatabase.use { db ->
                db.query("error_log", arrayOf("thread", "time", "sys", "model", "type", "msg"), "type = ?", arrayOf(type), null, null, "time desc", String.format("%s, %s", index, count)).use { cursor ->
                    var error: Error
                    while (cursor.moveToNext()) {
                        error = Error()
                        error.thread = cursor.getString(0)
                        error.time = cursor.getLong(1)
                        error.sys = cursor.getString(2)
                        error.model = cursor.getString(3)
                        error.type = cursor.getString(4)
                        error.msg = cursor.getString(5)
                        errors.add(error)
                    }
                }
            }
            return errors
        }

        /**
         * 获取日志
         * @return 日志集
         */
        fun getLogs(context: Context?, index: Int, count: Int): List<Error> {
            if (context == null) throw NullPointerException()
            check(!(index < 0 || count < 0))
            val errors: MutableList<Error> = ArrayList()
            ErrorDB(context.applicationContext).readableDatabase.use { db ->
                db.query("error_log", arrayOf("thread", "time", "sys", "model", "type", "msg"), null, null, null, null, "time desc", String.format("%s, %s", index, count)).use { cursor ->
                    var error: Error
                    while (cursor.moveToNext()) {
                        error = Error()
                        error.thread = cursor.getString(0)
                        error.time = cursor.getLong(1)
                        error.sys = cursor.getString(2)
                        error.model = cursor.getString(3)
                        error.type = cursor.getString(4)
                        error.msg = cursor.getString(5)
                        errors.add(error)
                    }
                }
            }
            return errors
        }

        /**
         * 获取所有日志
         * @return 日志集
         */
        fun getLogs(context: Context?): List<Error> {
            return getLogs(context, 0, Int.MAX_VALUE)
        }

        /**
         * 获取最新日志
         * @return 最新日志
         */
        fun getNewLog(context: Context?): Error? {
            if (context == null) throw NullPointerException()
            ErrorDB(context.applicationContext).readableDatabase.use { db ->
                db.query("error_log", arrayOf("thread", "time", "sys", "model", "type", "msg"), null, null, null, null, "time desc", "1").use { cursor ->
                    val error: Error
                    while (cursor.moveToNext()) {
                        error = Error()
                        error.thread = cursor.getString(0)
                        error.time = cursor.getLong(1)
                        error.sys = cursor.getString(2)
                        error.model = cursor.getString(3)
                        error.type = cursor.getString(4)
                        error.msg = cursor.getString(5)
                        return error
                    }
                }
            }
            return null
        }

        /**
         * 清空所有日志
         * @return 是否成功，false清空失败，true清空成功
         */
        fun clear(context: Context?) {
            if (context == null) throw NullPointerException()
            ErrorDB(context.applicationContext).readableDatabase.use { db -> db.delete("error_log", null, null) }
        }

        /**
         * 导出全部崩溃记录
         */
        @Throws(FileNotFoundException::class)
        fun export(context: Context?, path: File?) {
            val logs = getLogs(context)
            export(path, logs)
        }

        /**
         * 导出指定崩溃记录
         */
        @Throws(FileNotFoundException::class)
        fun export(path: File?, errors: List<Error>) {
            if (path == null) throw NullPointerException()
            if (!path.exists()) {
                if (!path.mkdirs()) {
                    throw FileNotFoundException()
                }
            }
            val sb = StringBuilder()
            sb.append(String.format("%s,%s,%s,%s,%s,%s\n", "thread", "time", "sys", "model", "type", "msg"))
            for (log in errors) {
                sb.append(String.format(Locale.CHINA, "%s,%tF %tT,%s,%s,%s,%s\n", log.thread, log.time, log.time, log.sys, log.model, log.type, log.msg!!.replace("\n", " ~ ")))
            }
            val now = System.currentTimeMillis()
            val file = File(path, String.format(Locale.CHINA, "%tF-%tT.csv", now, now))
            try {
                RandomAccessFile(file, "rw").use { out -> out.write(sb.toString().toByteArray(Charset.forName("GB2312"))) }
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

}